packages/components/src/internal/url/AppURLResolver.ts

```diff
@@ -132,74 +132,91 @@ export class AssayRunResolver implements AppRouteResolver {
+        const key = [containerPath, listIdNum].join('|');
+
+        if (isNaN(listIdNum) || !containerPath) {
             // skip it
-            return Promise.resolve(true);
-        } else if (this.lists.has(listIdNum)) {
+            return true;
+        } else if (this.lists.has(key)) {
             // resolve it
-            const newParts = [this.lists.get(listIdNum)];
-            return Promise.resolve(spliceURL(parts, newParts, listIdIndex));
+            const newParts = [this.lists.get(key)];
+            return spliceURL(parts, newParts, containerPathIndex, 2);
         } else if (this.fetched) {
             // skip it
-            return Promise.resolve(true);
-        } else {
-            // fetch it
-            return new Promise(resolve => {
-                return selectRows({
-                    schemaName: SCHEMAS.LIST_METADATA_TABLES.SCHEMA,
-                    queryName: SCHEMAS.LIST_METADATA_TABLES.LIST_MANAGER.queryName,
-                    columns: 'ListId,Name',
-                }).then(result => {
-                    this.fetched = true;
-
-                    // fulfill local cache
-                    const allLists = Map<number, string>().asMutable();
-                    const lists = result.models[result.key];
-                    for (var i in lists) {
-                        if (lists.hasOwnProperty(i)) {
-                            allLists.set(lists[i].ListId.value, lists[i].Name.value.toLowerCase());
-                        }
-                    }
-                    this.lists = allLists.asImmutable();
-
-                    // respond
-                    if (this.lists.has(listIdNum)) {
-                        // resolve it
-                        const newParts = [this.lists.get(listIdNum)];
-                        return resolve(spliceURL(parts, newParts, listIdIndex));
-                    }
-
-                    // skip it
-                    return resolve(true);
-                });
+            return true;
+        }
+
+        // fetch it
+        try {
+            const result = await selectRows({
+                schemaName: SCHEMAS.LIST_METADATA_TABLES.LIST_MANAGER.schemaName,
+                queryName: SCHEMAS.LIST_METADATA_TABLES.LIST_MANAGER.queryName,
+                columns: 'ListId,Name,Container/Path',
             });
+
+            this.fetched = true;
+
+            // fulfill local cache
+            this.lists = Object.values(result.models[result.key])
+                .reduce<Map<string, string>>((map, list) => {
+                    const _containerPath = caseInsensitive(list, 'Container/Path').value.toLowerCase();
+                    const _listId = caseInsensitive(list, 'ListId').value;
+                    const _name = caseInsensitive(list, 'Name').value.toLowerCase();
+
+                    const _key = [_containerPath, _listId].join('|');
+                    return map.set(_key, _name);
+                }, Map<string, string>().asMutable())
```

Comment:
You shouldn't need either the `asMutable()` or `asImmutable()` calls here since your reducer returns the result of `map.set`

Reply:
Yes, the difference is that with an immutable collection `map.set` will return the same instance as opposed to creating a new instance each time.